File: src/protocol/notifications.ts
/**
 * ECP Notification Constants
 *
 * All notification event names emitted by the ECP server (and client-side
 * synthetic events). Clients subscribe to these to receive real-time updates.
 *
 * Sources:
 *   Server emitters:  sendNotification / emitNotification / createNotification
 *   Client listeners: ecp.subscribe(...)
 */

export const Notifications = {
  // ── Authentication ────────────────────────────────────────────────────
  Auth: {
    /** Sent during WebSocket handshake when credentials are required. */
    Required: 'auth/required',
  },

  // ── Server lifecycle ──────────────────────────────────────────────────
  Server: {
    /** Sent to the client after a successful WebSocket connection. */
    Connected: 'server/connected',
  },

  // ── Client-side synthetic events ──────────────────────────────────────
  Ecp: {
    /** Emitted by the client ecpStore after a successful reconnection. */
    Reconnected: 'ecp/reconnected',
  },

  // ── File system ───────────────────────────────────────────────────────
  File: {
    DidChange: 'file/didChange',
    DidCreate: 'file/didCreate',
    DidDelete: 'file/didDelete',
  },

  // ── Terminal ──────────────────────────────────────────────────────────
  Terminal: {
    Output: 'terminal/output',
    Exit: 'terminal/exit',
    Title: 'terminal/title',
  },

  // ── Open document (editor buffer) ─────────────────────────────────────
  Document: {
    DidChange: 'document/didChange',
    DidChangeCursors: 'document/didChangeCursors',
    DidOpen: 'document/didOpen',
    DidClose: 'document/didClose',
  },

  // ── LSP ───────────────────────────────────────────────────────────────
  Lsp: {
    DidPublishDiagnostics: 'lsp/didPublishDiagnostics',
    ServerStatusChanged: 'lsp/serverStatusChanged',
  },

  // ── Theme ─────────────────────────────────────────────────────────────
  Theme: {
    Changed: 'theme/changed',
  },

  // ── Workspace ─────────────────────────────────────────────────────────
  Workspace: {
    DidChangeRoot: 'workspace/didChangeRoot',
  },

  // ── System prompt ─────────────────────────────────────────────────────
  SystemPrompt: {
    DidChange: 'systemPrompt/didChange',
  },

  // ── UI (built-in panel management) ────────────────────────────────────
  Ui: {
    SetLayout: 'ui/setLayout',
    AddPanel: 'ui/addPanel',
    RemovePanel: 'ui/removePanel',
    UpdatePanel: 'ui/updatePanel',
    LayoutSet: 'ui/layout/set',
    PresetsSwitch: 'ui/presets/switch',
  },

  // ── Layout (agent-driven layout requests) ─────────────────────────────
  Layout: {
    DidRequestAddTab: 'layout/didRequestAddTab',
    DidRequestSplit: 'layout/didRequestSplit',
    DidRequestFocus: 'layout/didRequestFocus',
    DidRequestOpenFile: 'layout/didRequestOpenFile',
    DidRequestPreset: 'layout/didRequestPreset',
    DidRequestClose: 'layout/didRequestClose',
    DidRequestCloseTab: 'layout/didRequestCloseTab',
    DidRequestActivateTab: 'layout/didRequestActivateTab',
    DidRequestState: 'layout/didRequestState',
  },

  // ── AI service ────────────────────────────────────────────────────────
  Ai: {
    StreamEvent: 'ai/stream/event',
    TodoUpdated: 'ai/todo/updated',

    // Dynamic session lifecycle (forwarded from ai service onSessionEvent)
    SessionCreated: 'ai/session_created',
    SessionDeleted: 'ai/session_deleted',
    SessionUpdated: 'ai/session_updated',
    MessageAdded: 'ai/message_added',

    Agent: {
      Handoff: 'ai/agent/handoff',
      Updated: 'ai/agent/updated',
      Joined: 'ai/agent/joined',
      Left: 'ai/agent/left',
    },

    Persona: {
      Created: 'ai/persona/created',
      Updated: 'ai/persona/updated',
      Deleted: 'ai/persona/deleted',
    },
  },

  // ── Agent service (standalone agent CRUD) ─────────────────────────────
  Agent: {
    Created: 'agent/created',
    Deleted: 'agent/deleted',
    Status: 'agent/status',
  },

  // ── Chat service ──────────────────────────────────────────────────────
  Chat: {
    Activity: 'chat/activity',

    Todo: {
      Replaced: 'chat/todo/replaced',
      Updated: 'chat/todo/updated',
      Deleted: 'chat/todo/deleted',
    },

    Document: {
      Created: 'chat/document/created',
      Updated: 'chat/document/updated',
      Deleted: 'chat/document/deleted',
    },
  },

  // ── Workflow service ──────────────────────────────────────────────────
  Workflow: {
    Created: 'workflow/created',
    Updated: 'workflow/updated',
    Deleted: 'workflow/deleted',
    DefaultChanged: 'workflow/defaultChanged',
    Activity: 'workflow/activity',
    Output: 'workflow/output',
    AwaitingInput: 'workflow/awaiting_input',
    SplitStarted: 'workflow/split/started',
    MergeCompleted: 'workflow/merge/completed',

    Execution: {
      Started: 'workflow/execution/started',
      Paused: 'workflow/execution/paused',
      Resumed: 'workflow/execution/resumed',
      Completed: 'workflow/execution/completed',
      Failed: 'workflow/execution/failed',
      Cancelled: 'workflow/execution/cancelled',
    },

    Message: {
      Created: 'workflow/message/created',
      Started: 'workflow/message/started',
      Delta: 'workflow/message/delta',
      Completed: 'workflow/message/completed',
      ToolUse: 'workflow/message/tool_use',
      Error: 'workflow/message/error',
    },

    Node: {
      Completed: 'workflow/node/completed',
    },

    Checkpoint: {
      Reached: 'workflow/checkpoint/reached',
      Responded: 'workflow/checkpoint/responded',
    },

    Permission: {
      Request: 'workflow/permission/request',
      Granted: 'workflow/permission/granted',
      Denied: 'workflow/permission/denied',
    },

    ToolCall: {
      Created: 'workflow/toolCall/created',
      Approved: 'workflow/toolCall/approved',
      Denied: 'workflow/toolCall/denied',
    },

    Tool: {
      Execution: 'workflow/tool/execution',
    },

    Agent: {
      Created: 'workflow/agent/created',
      Updated: 'workflow/agent/updated',
      Deleted: 'workflow/agent/deleted',
    },

    ReviewPanel: {
      Started: 'workflow/review_panel/started',
      Completed: 'workflow/review_panel/completed',
      Vote: 'workflow/review_panel/vote',
      Decision: 'workflow/review_panel/decision',
    },

    Context: {
      Added: 'workflow/context/added',
      Compacted: 'workflow/context/compacted',
      Expanded: 'workflow/context/expanded',
    },
  },
} as const;

// ── Utility type: union of all notification name strings ────────────────
type DeepValues<T> = T extends object ? DeepValues<T[keyof T]> : T;
export type NotificationName = DeepValues<typeof Notifications>;

File: rust/crates/ecp-protocol/src/notifications.rs
//! ECP notification event name constants.
//!
//! Notifications are server-to-client messages with no response expected.
//! Clients subscribe to these for real-time updates.

/// All ECP notification names, grouped by service namespace.
pub struct Notifications;

impl Notifications {
    // ── Authentication ──────────────────────────────────────────────────
    pub const AUTH_REQUIRED: &str = "auth/required";

    // ── Server lifecycle ────────────────────────────────────────────────
    pub const SERVER_CONNECTED: &str = "server/connected";

    // ── File system ─────────────────────────────────────────────────────
    pub const FILE_DID_CHANGE: &str = "file/didChange";
    pub const FILE_DID_CREATE: &str = "file/didCreate";
    pub const FILE_DID_DELETE: &str = "file/didDelete";

    // ── Terminal ────────────────────────────────────────────────────────
    pub const TERMINAL_OUTPUT: &str = "terminal/output";
    pub const TERMINAL_EXIT: &str = "terminal/exit";
    pub const TERMINAL_TITLE: &str = "terminal/title";

    // ── Document ────────────────────────────────────────────────────────
    pub const DOCUMENT_DID_CHANGE: &str = "document/didChange";
    pub const DOCUMENT_DID_CHANGE_CURSORS: &str = "document/didChangeCursors";
    pub const DOCUMENT_DID_OPEN: &str = "document/didOpen";
    pub const DOCUMENT_DID_CLOSE: &str = "document/didClose";

    // ── LSP ─────────────────────────────────────────────────────────────
    pub const LSP_DID_PUBLISH_DIAGNOSTICS: &str = "lsp/didPublishDiagnostics";
    pub const LSP_SERVER_STATUS_CHANGED: &str = "lsp/serverStatusChanged";

    // ── Theme ───────────────────────────────────────────────────────────
    pub const THEME_CHANGED: &str = "theme/changed";

    // ── Workspace ───────────────────────────────────────────────────────
    pub const WORKSPACE_DID_CHANGE_ROOT: &str = "workspace/didChangeRoot";

    // ── UI ──────────────────────────────────────────────────────────────
    pub const UI_SET_LAYOUT: &str = "ui/setLayout";
    pub const UI_ADD_PANEL: &str = "ui/addPanel";
    pub const UI_REMOVE_PANEL: &str = "ui/removePanel";
    pub const UI_UPDATE_PANEL: &str = "ui/updatePanel";

    // ── Layout ──────────────────────────────────────────────────────────
    pub const LAYOUT_DID_REQUEST_ADD_TAB: &str = "layout/didRequestAddTab";
    pub const LAYOUT_DID_REQUEST_SPLIT: &str = "layout/didRequestSplit";
    pub const LAYOUT_DID_REQUEST_FOCUS: &str = "layout/didRequestFocus";
    pub const LAYOUT_DID_REQUEST_OPEN_FILE: &str = "layout/didRequestOpenFile";
    pub const LAYOUT_DID_REQUEST_PRESET: &str = "layout/didRequestPreset";
    pub const LAYOUT_DID_REQUEST_CLOSE: &str = "layout/didRequestClose";

    // ── AI ──────────────────────────────────────────────────────────────
    pub const AI_STREAM_EVENT: &str = "ai/stream/event";
    pub const AI_TODO_UPDATED: &str = "ai/todo/updated";
    pub const AI_SESSION_CREATED: &str = "ai/session_created";
    pub const AI_SESSION_DELETED: &str = "ai/session_deleted";
    pub const AI_SESSION_UPDATED: &str = "ai/session_updated";
    pub const AI_MESSAGE_ADDED: &str = "ai/message_added";
    pub const AI_AGENT_HANDOFF: &str = "ai/agent/handoff";
    pub const AI_AGENT_UPDATED: &str = "ai/agent/updated";

    // ── Workflow ─────────────────────────────────────────────────────────
    pub const WORKFLOW_CREATED: &str = "workflow/created";
    pub const WORKFLOW_UPDATED: &str = "workflow/updated";
    pub const WORKFLOW_DELETED: &str = "workflow/deleted";
    pub const WORKFLOW_ACTIVITY: &str = "workflow/activity";
    pub const WORKFLOW_OUTPUT: &str = "workflow/output";
    pub const WORKFLOW_EXECUTION_STARTED: &str = "workflow/execution/started";
    pub const WORKFLOW_EXECUTION_PAUSED: &str = "workflow/execution/paused";
    pub const WORKFLOW_EXECUTION_COMPLETED: &str = "workflow/execution/completed";
    pub const WORKFLOW_EXECUTION_FAILED: &str = "workflow/execution/failed";
    pub const WORKFLOW_EXECUTION_CANCELLED: &str = "workflow/execution/cancelled";
    pub const WORKFLOW_MESSAGE_DELTA: &str = "workflow/message/delta";
    pub const WORKFLOW_MESSAGE_COMPLETED: &str = "workflow/message/completed";
    pub const WORKFLOW_CHECKPOINT_REACHED: &str = "workflow/checkpoint/reached";
    pub const WORKFLOW_PERMISSION_REQUEST: &str = "workflow/permission/request";
    pub const WORKFLOW_TOOL_EXECUTION: &str = "workflow/tool/execution";

    // ── Chat ────────────────────────────────────────────────────────────
    pub const CHAT_ACTIVITY: &str = "chat/activity";
    pub const CHAT_TODO_REPLACED: &str = "chat/todo/replaced";
    pub const CHAT_TODO_UPDATED: &str = "chat/todo/updated";
    pub const CHAT_DOCUMENT_CREATED: &str = "chat/document/created";
    pub const CHAT_DOCUMENT_UPDATED: &str = "chat/document/updated";

    // ── Agent ───────────────────────────────────────────────────────────
    pub const AGENT_CREATED: &str = "agent/created";
    pub const AGENT_DELETED: &str = "agent/deleted";
    pub const AGENT_STATUS: &str = "agent/status";
}

/// Type alias for notification names.
pub type NotificationName = &'static str;
